arg_scanner/lib/arg_scanner/state_tracker.rb

```diff
@@ -0,0 +1,145 @@
+require "set"
+require "tempfile"
+require "require_all"
+
+module ArgScanner
+  class StateTracker
+    def initialize
+      at_exit do
+        dir = ENV["ARG_SCANNER_STATE_TRACKER_DIR"]
+        dir = "." if dir.nil? || dir == ""
+        tmp_file = Tempfile.new(["classes-", ".json"], dir )
+        path = tmp_file.path
+        tmp_file.close!
+
+        begin
+          require_all Rails.root.join('lib')
+        rescue => e
+        end
+        begin
+          Rails.application.eager_load!
+        rescue => e
+        end
+
+        File.open(path,"w") { |file| print_json(file) }
+      end
+    end
+
+    private
+    def print_json(file)
+      file.puts("{")
+      print_load_path(file)
+      file.puts(",")
+      print_all_modules(file)
+      file.puts("}")
+    end
+
+    def get_constants_of_class(constants, parent, klass)
+      begin
+        constants.select {|const| parent.const_defined?(const)}.map do |const|
+          begin
+            parent.const_get(const)
+          rescue Exception => e
```

Comment:
What is being written down as mapped value if exception occurs? My concern is the next `select` block which nay throw NPE probably in these cases

Reply:
What do you mean? I don't get it.

Reply:
You're `map`ping elements to some other ones, getting new collection. Then, you select elements by predicate `const.is_a? klass`.
Imagine the case `parent.const_get(const)` throws Exception, and for example `nil` will be added to the collection. Then, on `select` phase you'll try to invoke `is_a?` on `nil` object which will fail.

UPD: oh, it's my Java thinking; `nil` is an Object, too, and has `is_a?` method, nevermind.